### ISSUE TITLE
DOI scripts have issues handling large amounts of items

### ISSUE THREAD

**Issue by KevinVdV** [DSpace/DSpace#8228]
The DOI scripts from DSpace 6 onwarths try to process all DOI's in a certain state in one go. This can lead to issues if you are doing large imports & want to bulk send DOI's to Datacite.

The biggest issue is with the actual registration, what the code does then:
* Loops over all DOI's that are ready to be registered
* Sends the data to Datacite
* Adds the DOI that it retrieved from Datacite to the item metadata
* Updates the DOI row to reflect that the DOI was registered
* uncached the item & proceeds in the loop
* Once the loop is done a single commit happens that will run all the consumers (which could load all items back into memory)
* Commits the database

The problem comes when processing the consumers as this could load up all the items into the memory again & if it fails here the metadata & DOI row change aren't written to the database & there is nothing for it but to run the script again.

I see two options to remedy this issue:

* We rewrite the code so that it doesn't loop over all the DOI's in a single loop, but rather paces itself (1000 at a time for example)
* We put a limit to the maximum number of DOI's that can be processed in a single run of the script (can be a CLI parameter as we do for filter media). The theory being if this number is higher then the daily added amount the DOI script will eventually catch up.

**Comment by mwoodiupui** [comment]
There is a third way, but it may take more work.  We should have one Session running a single read-only, forward-only query to produce the stream of DOIs to register, and a second Session to update and commit for each DOI.  Then we shouldn't need to do any manual batching or manual decaching; Hibernate should handle these as it knows best.  The unit of atomicity here is a single DOI -- we don't need to hold a transaction open for a thousand.

The difficulty with this approach is that the persistence layer is designed for a single persistence session per request.  (The request here is a run of doi-organiser.)  The opportunity is that we have a number of bulk update operations that should work this way.

**Comment by tdonohue** [comment]
Definitely sounds like a bug here & I like the suggestions. I do agree that @mwoodiupui 's approach should be looked at...it's the "ideal" behavior here, but not sure if other performance issues will arise.  If so, a fallback to one of the approaches mentioned by @KevinVdV seems reasonable.

As DOIs are not used by most sites at this time, I'm placing this in "Tier 3: Medium Priority" for now. That said, if anyone wants to work on this earlier, volunteer work always welcome

**Comment by bram-atmire** [comment]
Duplicate of https://github.com/DSpace/DSpace/issues/9622
